libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.kt

```diff
@@ -0,0 +1,1376 @@
+import com.salesforce.androidsdk.auth.HttpAccess.DEFAULT
+import com.salesforce.androidsdk.auth.OAuth2.IdServiceResponse
+import com.salesforce.androidsdk.auth.OAuth2.OAuthFailedException
+import com.salesforce.androidsdk.auth.OAuth2.TokenEndpointResponse
+import com.salesforce.androidsdk.auth.OAuth2.addAuthorizationHeader
+import com.salesforce.androidsdk.auth.OAuth2.callIdentityService
+import com.salesforce.androidsdk.auth.OAuth2.exchangeCode
+import com.salesforce.androidsdk.auth.OAuth2.getAuthorizationUrl
+import com.salesforce.androidsdk.auth.OAuth2.getFrontdoorUrl
+import com.salesforce.androidsdk.auth.OAuth2.revokeRefreshToken
+import com.salesforce.androidsdk.auth.OAuth2.swapJWTForTokens
+import com.salesforce.androidsdk.config.RuntimeConfig.getRuntimeConfig
+import com.salesforce.androidsdk.push.PushMessaging.register
+import com.salesforce.androidsdk.rest.ClientManager
+import com.salesforce.androidsdk.rest.ClientManager.LoginOptions
+import com.salesforce.androidsdk.rest.RestClient.clearCaches
+import com.salesforce.androidsdk.security.BiometricAuthenticationManager
+import com.salesforce.androidsdk.security.BiometricAuthenticationManager.Companion.isEnabled
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator.getRandom128ByteKey
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator.getSHA256Hash
+import com.salesforce.androidsdk.ui.LoginActivity.Companion.PICK_SERVER_REQUEST_CODE
+import com.salesforce.androidsdk.ui.OAuthWebviewHelper.AccountOptions.Companion.fromBundle
+import com.salesforce.androidsdk.util.EventsObservable
+import com.salesforce.androidsdk.util.EventsObservable.EventType.AuthWebViewPageFinished
+import com.salesforce.androidsdk.util.MapUtil
+import com.salesforce.androidsdk.util.MapUtil.addBundleToMap
+import com.salesforce.androidsdk.util.SalesforceSDKLogger.d
+import com.salesforce.androidsdk.util.SalesforceSDKLogger.e
+import com.salesforce.androidsdk.util.SalesforceSDKLogger.w
+import com.salesforce.androidsdk.util.UriFragmentParser
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.Default
+import kotlinx.coroutines.Dispatchers.Main
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import okhttp3.Request.Builder
+import org.json.JSONArray
+import org.json.JSONObject
+import java.lang.String.format
+import java.net.URI
+import java.net.URI.create
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import java.util.function.Consumer
+
+/**
+ * A helper class to manage a web view going through the OAuth login process.
+ * The basic flow is:
+ * a) Load and show the login page to the user
+ * b) User login and app authorization
+ * c) Navigate to the authentication completion URL and token fetch
+ * d) Call the id service to obtain additional info about the user
+ * e) Create a local account and return an authentication result bundle
+ */
+open class OAuthWebviewHelper : KeyChainAliasCallback {
```

Comment:
Just like the other Kotlin conversions, I preserved the structure, logic and API here entirely.  In this case, there may be a lot we could do in the future to modernize this class and the authentication flow, but I have a future work item open for that.